pkg/hs/urbit-king/lib/Urbit/Vere/Ames/Packet.hs

```diff
@@ -0,0 +1,102 @@
+{-|
+  Parsing of Ames packets
+-}
+
+module Urbit.Vere.Ames.Packet where
+
+import Urbit.Prelude
+
+import Data.Bits
+import Data.LargeWord
+import Data.Serialize
+
+import Urbit.Arvo (AmesDest)
+import Urbit.Noun.Tree (mug)
+
+data Packet = Packet
+  { pktVersion    :: Word8
+  , pktEncrypted  :: Bool
+  --
+  , pktSndr       :: Ship
+  , pktRcvr       :: Ship
+  , pktOrigin     :: Maybe AmesDest
+  , pktContent    :: Bytes
+  }
+
+instance Show Packet where
+  show Packet {..}
+    = "Packet {pktVersion = "
+   <> show pktVersion
+   <> ", pktEncrypted = "
+   <> show pktEncrypted
+   <> ", pktSndr = "
+   <> show pktSndr
+   <> ", pktRcvr = "
+   <> show pktRcvr
+   <> ", pktOrigin = "
+   <> show pktOrigin
+   <> ", pktContent = "
+   <> showUD (bytesAtom $ unBytes pktContent)
+   <> "}"
+
+instance Serialize Packet where
+  get = do
+    -- header
+    head <- getWord32le
+    let pktVersion   =         head    .&. 0b111        & fromIntegral
+    let checksum     = shiftR  head  3 .&. (2 ^ 20 - 1)
+    let sndrRank     = shiftR  head 23 .&. 0b11
+    let rcvrRank     = shiftR  head 25 .&. 0b11
+    let pktEncrypted = testBit head 27  &  not  -- loobean
+    -- verify checksum
+    lookAhead $ do
+      len  <- remaining
+      body <- getBytes len
+      -- XX mug (marked "TODO") is implemented as "slowMug" in U.N.Tree. Ominous
+      -- Also, toNoun will copy the bytes into an atom. We probably want a mugBS
+      let chk = fromIntegral (mug $ toNoun $ MkBytes body) .&. (2 ^ 20 - 1)
```

Comment:
So, is this super slow right now, or what? Once you're reasonably confident in this as a whole, you should try putting it on ~nus and see if it can keep up.

Reply:
On reflection, I think that mug is only appreciably slow for nonatomic nouns, but I could be wrong about this. I think that you propose a good test. How do we track whether it is keeping up?

Reply:
~nus was dropping packets without stateless forwarding. If it's not getting pressured into dropping packets when running stateless forwarding (I see kh has printfs for this too), then it's keeping up. (: